changelog/ob.md

```diff
@@ -0,0 +1,282 @@
+by OB. Whether a pointer is allocate memory using the GC or some other
+storage allocator is immaterial to OB, they are not distinguished and are
+handled identically.
+
+The system is only active in functions annotated with the `@live` attribute.
+It is applied after semantic processing is done as purely a check
+for violations of the OB rules. No new syntax is added. No change is made
+to the code generated.
+If `@live` functions
+call non-`@live` functions, those called functions are expected to present
+an `@live` compatible interface, although it is not checked.
+
+The OB system will detect as errors:
+
+* dereferencing pointers that are in an invalid state
+* more than one active pointer to a mutable memory object
+
+It will not detect attempts to dereference `null` pointers or possibly
+`null` pointers. This is unworkable because there is no current method
+of annotating a type as a non-`null` pointer.
+
+
+## Core OB Principle
+
+The OB design follows from the following principle:
+
+For each memory object, there can exist either exactly one mutating pointer
+to it, or multiple non-mutating (read-only) pointers.
+
+
+## Design
+
+The single mutating pointer is called the "owner" of the memory object.
+It transitively owns the memory object and all memory objects accessible
+from it (i.e. the memory object graph). Since it is the sole pointer to
+that memory object, it can safely
+manage the memory (change its shape, allocate, free and resize) without
+pulling the rug out from under any other pointers (mutating or not)
+that may point to it.
+
+If there are multiple read-only pointers to the memory object graph,
+they can safely read from it without being concerned about the memory
+object graph being changed underfoot.
+
+The rest of the design is concerned with how pointers become owners,
+read only pointers, and invalid pointers, and how the Core OB Principle
+is maintained at all times.
+
+
+### Tracked Pointers
+
+The only pointers that are tracked are those declared in the `@live` function
+as `this`, function parameters or local variables. Variables from other
+functions are not tracked, even `@live` ones, as the analysis of interactions
+with other functions depends
+entirely on that function signature, not its internals.
+Parameters that are const are not tracked.
+
+
+### Pointer States
+
+Each pointer is in one of the following states:
+
+1. Undefined
+
+The pointer is in an invalid state. Dereferencing such a pointer is
+an error.
+
+2. Owner
+
+The owner is the sole pointer to a memory object graph.
+An Owner pointer must not be `scope` and must be a pointer to mutable.
+
+3. Borrowed
+
+A Borrowed pointer is one that temporarily becomes the sole pointer
+to a memory object graph. It enters that state via assignment
+from an owner pointer, and the owner then enters the Lent state
+until after the last use of the borrowed pointer.
+
+A Borrowed pointer must have the `scope` attribute and must
```

Comment:
If such a pointer isn't `scope`, what happens? Is that forbidden by the design?

Reply:
If it isn't `scope`, then it is a transfer of the source pointer to the target pointer. The state of the source pointer then becomes Undefined.

Reply:
I think that should be clarified in the description.

Reply:
I've rewritten parts of ob.md.